src/service/fixed-layer.js

```diff
@@ -0,0 +1,417 @@
+    /** @const {!Document} */
+    this.doc = doc;
+
+    /** @private @const {!Vsync} */
+    this.vsync_ = vsync;
+
+    /** @private {number} */
+    this.paddingTop_ = paddingTop;
+
+    /** @private @const {boolean} */
+    this.transfer_ = transfer;
+
+    /** @private {?Element} */
+    this.fixedLayer_ = null;
+
+    /** @const @private {!Array<!FixedElementDef>} */
+    this.fixedElements_ = [];
+  }
+
+  /**
+   * @param {boolean} visible
+   */
+  setVisible(visible) {
+    if (this.fixedLayer_) {
+      this.vsync_.mutate(() => {
+        setStyle(this.fixedLayer_, 'visibility',
+            visible ? 'visible' : 'hidden');
+      });
+    }
+  }
+
+  /**
+   * Must be always called after DOMReady.
+   */
+  setup() {
+    // Find all `position:fixed` elements.
+    const stylesheets = document.styleSheets;
+    const fixedSelectors = [];
+    for (let i = 0; i < stylesheets.length; i++) {
+      const stylesheet = stylesheets[i];
+      if (stylesheet.disabled ||
+              !stylesheet.ownerNode ||
+              stylesheet.ownerNode.tagName != 'STYLE' ||
+              stylesheet.ownerNode.hasAttribute('amp-boilerplate') ||
+              stylesheet.ownerNode.hasAttribute('amp-runtime') ||
+              stylesheet.ownerNode.hasAttribute('amp-extension')) {
+        continue;
+      }
+      this.discoverFixedSelectors_(stylesheet.cssRules, fixedSelectors);
+    }
+
+    try {
+      const idMap = {};
+      fixedSelectors.forEach(selector => {
+        const elements = this.doc.querySelectorAll(selector);
+        for (let i = 0; i < elements.length; i++) {
+          if (i > 10) {
+            // We shouldn't have too many of `fixed` elements.
+            break;
+          }
+          this.setupFixedElement_(elements[i], selector, idMap);
+        }
+      });
+    } catch (e) {
+      // Fail quietly.
+      setTimeout(() => {throw e;});
+    }
+
+    // Sort in document order.
+    this.fixedElements_.sort(function(fe1, fe2) {
+      // 8 | 2 = 0x0A
+      // 2 - preceeding
+      // 8 - contains
+      if (fe1.element.compareDocumentPosition(fe2.element) & 0x0A != 0) {
+        return -1;
+      }
+      return 1;
+    });
+
+    this.update();
+  }
+
+  /**
+   * @param {number} paddingTop
+   */
+  updatePaddingTop(paddingTop) {
+    this.paddingTop_ = paddingTop;
+    this.update();
+  }
+
+  /**
+   * Performs fixed actions.
+   * 1. Updates `top` styling if necessary.
+   * 2. On iOS/Iframe moves elements between fixed layer and BODY depending on
+   * whether they are currently visible and fixed.
+   * @return {!Promise}
+   */
+  update() {
+    if (this.fixedElements_.length == 0) {
+      return Promise.resolve();
+    }
+
+    // Some of the elements may no longer be in DOM.
+    /** @type {!Array<number>} */
+    const toRemove = [];
+    this.fixedElements_.forEach((fe, i) => {
+      if (!this.doc.contains(fe.element)) {
+        toRemove.push(i);
+      }
+    });
+    if (toRemove.length > 0) {
```

Comment:
dont think you need this, forEach just wouldnt do anything with an empty array anyways right?